### PR TITLE
[mmu probing] pr12.fix: Common bug fixes for physical platform validation

### LINKED ISSUE #22605: Enhancement: Enable MMU threshold probing test on Broadcom Tomahawk 2 platforms
### Is it platform specific

broadcom

### Importance or Severity

Medium

### Description of the enhancement

MMU threshold probing functionality tests needs to be validated on Broadcom Tomahawk 2 platforms to ensure there are no corner cases and the test can run successfully.

### Current Behavior

Have not completed joint debugging for MMU threshold probing on Broadcom Tomahawk 2 platforms.

### Proposed Behavior

Enable and validate MMU threshold probing functionality on Broadcom Tomahawk 2 platforms (hwsku: Arista-7260CX3-C64, Arista-7260CX3-D108C8, Arista-7260CX3-D108C10).

Target SKUs:
- Arista-7260CX3-C64
- Arista-7260CX3-D108C8
- Arista-7260CX3-D108C10

### PR DESCRIPTION
**Pull request by XuChen-MSFT**
### Description of PR

Fix three common bugs discovered during physical platform validation, plus add missing mark_conditions skip entries for the MMU threshold probing framework.

1. **Dynamic dut/asic index in probe setup_traffic** -- On dualtor topologies, `test_port_ips` keys may not start from 0. All three probing files used hardcoded `[0][0]` causing `KeyError`. Fixed with `next(iter(...))`.

2. **Dynamic key lookup for testPortIds/testPortIps** -- `testPortIds[0]` assumed dict keys starting from 0. On multi-DUT/ASIC configs, the first key may differ. Fixed with dynamic key lookup.

3. **Retry with observability for rx_port_resolver** -- On T1 topologies, BGP routes may not be converged at test start, causing `get_rx_port()` to return None. Added retry logic (3 attempts, 5s interval) with logging.

4. **Add test_qos_probe.py to mark_conditions skip for VS multi-asic-t1 and T2** -- `test_qos_probe.py` (added in PR09 #22547) was missing from the conditional mark skip lists for VS virtual chassis topologies. `test_qos_sai.py` already has these skip entries because SAI thrift-based QoS tests do not support cross-ASIC probing on virtual switch platforms. Apply the same skip conditions for consistency.

### Type of change

- [x] Bug fix

#### What is the motivation for this PR?

During physical platform validation (pr01-pr11), bugs 1-3 were discovered on real testbeds. Bug 4 was found during PR CI: `test_qos_probe.py` errored on multi-asic-t1 and T2 KVM topologies because it was missing the same `mark_conditions` skip entries that `test_qos_sai.py` already has.

#### How did you verify/test it?

Validated on physical testbeds via ElasticTest pipeline (branch `xuchen3/internal/pr-phy-test-r11`):

| ASIC | Topology | Result | Test Plan |
|------|----------|--------|-----------|
| TH2 (7260CX3) | t0-116 | PASS (22p/0f) | [69da5889](https://elastictest.org/scheduler/testplan/69da588964259d7df8aa6a59) |
| TH2 (7260CX3) | t0-118 | PASS (22p/0f) | [69d95fc9](https://elastictest.org/scheduler/testplan/69d95fc9606c54b3249a0a1b) |
| TH2 (7260CX3) | t1-64-lag | PASS (21p/0f) | [69da588d](https://elastictest.org/scheduler/testplan/69da588d64259d7df8aa6a58) |
| TH2 (7260CX3) | dualtor-aa-56 | PASS (24p/0f/46s) | [69da6208](https://elastictest.org/scheduler/testplan/69da620859cbd7d94a4fe696) |
| TH (7060CX) | t0 | PASS (22p/0f/43s) | [69e18653](https://elastictest.org/scheduler/testplan/69e186536e6dd6a80874c6e2) |
| TH5 (7060X6) | t0-isolated | PASS (20p/0f/45s) | [69d8c3e0](https://elastictest.org/scheduler/testplan/69d8c3e059cbd7d94a4fe4ab) |
| TH5 (7060X6) | t1-isolated | PASS (23p/1f/86s) | [69d964dd](https://elastictest.org/scheduler/testplan/69d964dd6906fc5b3eb65a58) |
| TD3 (7050CX3) | t0 | PASS (5p/0f/40s) | [69d76224](https://elastictest.org/scheduler/testplan/69d762246906fc5b3eb6570d) |

### Merge Dependency

Depends on (all merged): #22541, #22547, #23341

### Related Issues

Fixes #22605 (TH2), contributes to #22604 (TH), #22603 (TD3), #22606 (TH5)

### Documentation

Related PRs: #22547 (pr09), #23341 (pr10), #23649 (pr11)

### CODE REVIEW

File: tests/common/plugins/conditional_mark/tests_mark_conditions_vs_t1_multiasic.yaml [hunk 1]
```diff
@@ -543,6 +543,11 @@ qos/test_qos_masic.py:
     conditions:
     - asic_type in ['vs'] and 't1-8-lag' in topo_name
     reason: This test case either cannot pass or should be skipped on virtual chassis
+qos/test_qos_probe.py:
+  skip:
+    conditions:
+    - asic_type in ['vs'] and 't1-8-lag' in topo_name
+    reason: This test case either cannot pass or should be skipped on virtual chassis
```
Comment (StormLiangMS) [inline tests/common/plugins/conditional_mark/tests_mark_conditions_vs_t1_multiasic.yaml:550]: hi @XuChen-MSFT  for skip, if we plan to support this later on, pls file a issue to skip for now. I'd like to merge this to unblock, but pls have a separate PR for this.

File: tests/common/plugins/conditional_mark/tests_mark_conditions_vs_t2.yaml [hunk 2]
```diff
@@ -800,6 +800,11 @@ qos/test_oq_watchdog.py:
     conditions:
     - asic_type in ['vs'] and 't2' in topo_name
     reason: This test case either cannot pass or should be skipped on virtual chassis
+qos/test_qos_probe.py:
+  skip:
+    conditions:
+    - asic_type in ['vs'] and 't2' in topo_name
+    reason: This test case either cannot pass or should be skipped on virtual chassis
```
Comment (StormLiangMS) [inline tests/common/plugins/conditional_mark/tests_mark_conditions_vs_t2.yaml:807]: same. @XuChen-MSFT

**Review by StormLiangMS — APPROVED** [review]
LGTM

**Comment by StormLiangMS** [comment]
## Review: PR #24024 — `[mmu probing] pr12.fix: Common bug fixes for physical platform validation`

**4 commits, ~7 files** | Well-structured with clean 1-fix-per-commit separation and thorough validation across 8 testbed configs.

---

### ✅ What looks good

1. **Dynamic dut/asic index (Fixes 1 & 2)** — `next(iter(...))` is correct and Pythonic for dict key fallback. The hardcoded `[0][0]` was a real bug on dualtor/multi-DUT topologies.

2. **mark_conditions (Fix 4)** — Mirrors existing `test_qos_sai.py` skip entries. Correct and consistent.

3. **Validation** — Excellent coverage: TH, TH2, TH5, TD3 across t0, t1, dualtor topologies.

---

### ⚠️ Minor suggestions

1. **Fix 2 (`__buildTestPorts`)** — The fallback silently overrides both src *and* dst indices when `src_dut_idx` is not found:
   ```python
   if src_dut_idx not in testPortIds:
       src_dut_idx = next(iter(testPortIds))
       dst_dut_idx = src_dut_idx  # dst forced to same as src
   ```
   Consider adding a log warning when fallback activates, so unexpected topology mismatches don't fail silently.

2. **Fix 3 (retry for `rx_port_resolver`)** — `from probing_observer import ProbingObserver` is inside the method body. Consider moving to file-level imports. Also verify that `get_rx_port` actually raises `AssertionError` (not returning `None`), otherwise the retry won't trigger.

3. **No unit tests** for retry logic or dynamic key fallback — understandable for test infra, but worth noting.

---

### Summary

| Aspect | Rating |
|--------|--------|
| Correctness | ✅ Fixes real bugs found on physical testbeds |
| Code clarity | ✅ Good comments explaining "why" |
| Risk | Low — test infra only |

**Overall: LGTM** with the minor suggestions above.

File: tests/common/plugins/conditional_mark/tests_mark_conditions_vs_t1_multiasic.yaml (hunk 1 above)
Comment (XuChen-MSFT) [reply tests/common/plugins/conditional_mark/tests_mark_conditions_vs_t1_multiasic.yaml:550]: track future fix in this issue https://github.com/sonic-net/sonic-mgmt/issues/24051

File: tests/common/plugins/conditional_mark/tests_mark_conditions_vs_t2.yaml (hunk 2 above)
Comment (XuChen-MSFT) [reply tests/common/plugins/conditional_mark/tests_mark_conditions_vs_t2.yaml:807]: track future fix in this issue https://github.com/sonic-net/sonic-mgmt/issues/24051
